tapdb/addrs_test.go

```diff
@@ -172,6 +172,9 @@ func TestAddressInsertion(t *testing.T) {
 	}
 	require.NoError(t, addrBook.InsertAddrs(ctx, addrs...))
 
+	// Now insert the addresses again, to test that upsert works as well.
+	require.NoError(t, addrBook.InsertAddrs(ctx, addrs...))
```

Comment:
What's your exact use case here? Because I'm not sure what the user experience of this change will be. Since we won't update any fields of the address, we will silently ignore any intended changes by the user.
Do you perhaps just need a `QueryAddrByKey` RPC or something along those lines?

Reply:
An address that is created with a given amount, script and internal keys, etc would need to be stored by the caller of `NewAddr` and fetched with `QueryAddrs` if we expect the caller application to restart at any point in time. This means that we'd need to guess when the address was created and scan through the results of `QueryAddrs`.

Maybe instead of adding `QueryAddrByKey` (which imo would be useful), we could make upsertion such that it will not change any of the fields if the address is already present. wdyt?

Reply:
Ah, okay. So how about we add a check that compares all fields of the upserted address to the one provided in the RPC request of `NewAddr`. If all fields are equal, just return the address.
If any field is different (e.g. the user expects to be able to change the proof courier address for example), we'd return an error with a message like "address with given key already exists with different properties"?

That way we can serve your use case but also not have confusing/silent behavior that the user might not understand.

Reply:
Done, PTAL!